Extension/src/LanguageServer/settings.ts

```diff
@@ -996,28 +919,99 @@ export class OtherSettings {
+            this.logValidationError(sectionName, settingName, "no default value");
+            return defaultValue;
+        }
+        return setting.defaultValue;
+    }
 
-    public get customTextMateRules(): TextMateRule[] | undefined { return vscode.workspace.getConfiguration("editor.tokenColorCustomizations").get<TextMateRule[]>("textMateRules"); }
-    public getCustomThemeSpecificTextMateRules(themeName: string): TextMateRule[] | undefined { return vscode.workspace.getConfiguration(`editor.tokenColorCustomizations.[${themeName}]`, this.resource).get<TextMateRule[]>("textMateRules"); }
+    private getAsBoolean(sectionName: string, settingName: string, resource: any, defaultValue: boolean): boolean {
+        const section = vscode.workspace.getConfiguration(sectionName, resource);
+        const value = section.get<any>(settingName);
+        if (isBoolean(value)) {
+            return value;
+        }
+        const setting = section.inspect<any>(settingName);
+        if (setting?.defaultValue === undefined || setting.defaultValue === null) {
+            this.logValidationError(sectionName, settingName, "no default value");
+            return defaultValue;
+        }
+        return setting.defaultValue;
+    }
+
+    private getAsNumber(sectionName: string, settingName: string, resource: any, defaultValue: number, minimum?: number, maximum?: number): number {
+        const section = vscode.workspace.getConfiguration(sectionName, resource);
+        const value = section.get<any>(settingName);
+        // Validates the value is a number and clamps it to the specified range. Allows for undefined maximum or minimum values.
+        if (isNumber(value)) {
+            if (minimum !== undefined && value < minimum) {
+                return minimum;
+            }
+            if (maximum !== undefined && value > maximum) {
+                return maximum;
+            }
+            return value;
+        }
+        const setting = section.inspect<any>(settingName);
+        if (setting?.defaultValue === undefined || setting.defaultValue === null) {
+            this.logValidationError(sectionName, settingName, "no default value");
+            return defaultValue;
+        }
+        return setting.defaultValue;
+    }
+
+    private getAsAssociations(sectionName: string, settingName: string, resource?: any): Associations {
+        const section = vscode.workspace.getConfiguration(sectionName, resource);
+        const value = section.get<any>(settingName);
+        if (isValidMapping(value, isString, isString)) {
+            return value as Associations;
+        }
+        const setting = section.inspect<any>(settingName);
+        if (setting?.defaultValue === undefined || setting.defaultValue === null) {
+            this.logValidationError(sectionName, settingName, "no default value");
+        }
+        return setting?.defaultValue as Associations;
```

Comment:
The `setting?.defaultValue as Associations;` could possibly be undefined, but the return type is not typed t include `undefined`.